test/unit/kubernetes-deploy/deploy_task_test.rb

```diff
@@ -26,22 +26,83 @@ def test_initializer
     assert_logs_match(/Template directory (\S+) doesn't exist/)
   end
 
+  def test_resource_deploy_order_cycle
+    task = deploy_task
+
+    list = []
+    list << resource_class(kind: 'a', deps: %w(b))
+    list << resource_class(kind: 'b', deps: %w(c))
+    list << resource_class(kind: 'c', deps: %w(a))
+
+    task.stubs(:all_resources).returns(list)
+
+    assert_raises(KubernetesDeploy::FatalDeploymentError) do
+      task._build_predeploy_sequence
+    end
+  end
+
+  def test_resource_deploy_order_correcntess
+    task = deploy_task
+
+    list = []
+    list << resource_class(kind: 'a', deps: %w(b k))
+    list << resource_class(kind: 'b', deps: [])
+    list << resource_class(kind: 'c', deps: %w(h i d))
+    list << resource_class(kind: 'd', deps: %w(j h))
+    list << resource_class(kind: 'e', deps: [])
+    list << resource_class(kind: 'f', deps: %w(e))
+    list << resource_class(kind: 'g', deps: [])
+    list << resource_class(kind: 'h', deps: [])
+    list << resource_class(kind: 'i', deps: %w(j))
+    list << resource_class(kind: 'j', deps: %w(g))
+    list << resource_class(kind: 'k', deps: %w(i))
+
+    task.stubs(:all_resources).returns(list)
+    order = task._build_predeploy_sequence
+
+    # We have the right number of resources
+    assert_equal list.count, order.count
+
+    # All resource should be in the list
+    list.each { |r| assert_includes order, r.kind }
+
+    # All preconditions should be respected
+    order.each_with_index do |r, idx|
+      klass = list.detect { |e| e.kind == r }
+      klass.predeploy_dependencies.each do |dep|
+        pos = order.map.with_index { |e, i| e == dep ? i : nil }.compact.first
+        # The current resource is deployed *after* its deps
+        assert_operator idx, :>, pos, "#{r} requires #{dep} but got #{order}"
```

Comment:
I did not want to hardcode a particular total order here because that's not important. What's important is that the result is in any valid toplogical sort order, which may not be unique.